src/phoenix/db/insertion/types.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+from copy import copy
+from dataclasses import dataclass, field
+from datetime import datetime, timezone
+from typing import (
+    AsyncContextManager,
+    Callable,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+)
+
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from phoenix.db import models
+from phoenix.db.helpers import SupportedSQLDialect
+from phoenix.db.insertion.constants import DEFAULT_RETRY_ALLOWANCE, DEFAULT_RETRY_DELAY_SEC
+from phoenix.db.insertion.helpers import as_kv, insert_on_conflict
+
+logger = logging.getLogger("__name__")
+
+
+class Insertable(Protocol):
+    @property
+    def row(self) -> models.Base: ...
+
+
+_AnyT = TypeVar("_AnyT")
+_PrecursorT = TypeVar("_PrecursorT")
+_InsertableT = TypeVar("_InsertableT", bound=Insertable)
+_RowT = TypeVar("_RowT", bound=models.Base, covariant=True)
+
+
+@dataclass(frozen=True)
+class _Wrapper(Generic[_AnyT]):
```

Comment:
This is pretty generic name. Maybe just `_Iterable`? Not sure it really needs to be a separate class if it's only used once, though.

Reply:
good point. it's revised